[Silas Tay] iP

SilasTSL opened the pull request:
# Luffy

> Relax, allow the mind to become empty, and surprise yourself with the great treasure that begins to flow from your soul. -Someone cool

Luffy helps you keep track of your tasks in a quick and neat manner. It includes:
- Text-based input
- Simple to pick up
- Super fast to use

All you need to do to use it is:
1. Download it [Here](https://github.com/SilasTSL/ip)
2. Double click it
3. Add your tasks
4. Enjoy!

Here's the `main` method for Luffy:
```
    /**
     * Main method for Luffy.
     *
     * @param args User commandline input
     */
    public static void main(String[] args) {
        new Luffy(FILE_PATH).run();
    }
```

Checklist of things to do with Luffy:
- [x] Download Luffy
- [ ] Try adding Todos, Events and Deadlines
- [ ] Deleting Todos, Events and Deadlines

Best thing is, Luffy is **FREE**!

Enjoy :+1:!

Commit:
Add Gradle support

SilasTSL committed:
Changed logo and file name to Luffy

SilasTSL committed:
Added echo

SilasTSL committed:
Added list

SilasTSL committed:
Added mark/unmark, Task class

SilasTSL committed:
Added todo, deadline, event

SilasTSL committed:
Changed EXPECTED.TXT and test.sh

SilasTSL committed:
Handle incorrect inputs

SilasTSL committed:
Added delete

SilasTSL committed:
Used Java collections

SilasTSL committed:
Added Javadocs

SilasTSL committed:
Javadocs for constructors

SilasTSL committed:
Added local save file

SilasTSL committed:
Used java DateTime

SilasTSL committed:
Merge branch 'branch-Level-7'
Level-7 commit

SilasTSL committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Luffy.java

SilasTSL committed:
Further OOP

SilasTSL committed:
Created package luffy

SilasTSL committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

SilasTSL committed:
Added gradle

SilasTSL committed:
Added JUnit testing

SilasTSL committed:
Fixed storage to create data folder

SilasTSL committed:
Added and fixed JavaDocs

SilasTSL committed:
Fixing JavaDocs

SilasTSL committed:
Fixing Coding Standards

SilasTSL committed:
Added find functionality

SilasTSL committed:
Revert "Added find functionality"

This reverts commit 01aa2b36df873e487347c406af7e9e2735f9e5e9.

SilasTSL committed:
Added find functionality

SilasTSL committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/luffy/Luffy.java

SilasTSL committed:
Merge branch 'branch-A-CodingStandard'
Merged Coding Standard branch to master.

SilasTSL committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/luffy/TaskList.java

SilasTSL committed:
Fix JavaDoc spacings

desmondyst commented on src/main/java/luffy/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+    public Deadline(String description, String by) {
+        super(description);
+        this.type = "[D]";
+        this.by = by;
```
Any reason why you decided to put the `by` attribute as part of `Task` class? Perhaps it will be better to have it as an attribute of `Deadline` class as it is only applicable to `Deadline` class and not other class that inherit from `Task`. Same goes for other fields such as `at`.

desmondyst commented on src/main/java/luffy/Luffy.java:
```diff
@@ -0,0 +1,68 @@
+            this.tasks = new TaskList(storage.load());
+        } catch (DukeException e) {
+            this.ui.showLoadingError();
+            this.tasks = new TaskList();
```
Perhaps can come up with alternative to handle the logic here? In general, not the best practice to use `try-catch` block for control flow.

desmondyst commented on src/main/java/luffy/Storage.java:
```diff
@@ -0,0 +1,102 @@
+                    String[] taskLine = scanner.nextLine().split(Pattern.quote(" | "));
+                    String taskDescription = taskLine[2];
+                    switch (taskLine[0]) {
+                        case "[T]":
```
Just some small points on styling.
```suggestion
                    case "[T]":
```

desmondyst replied:
Perhaps u can try to make use of `ENUMS`? As quoted from the textbook - "Use enumerations when a certain variable can take only a small number of finite values."

desmondyst commented on src/main/java/luffy/Task.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public abstract class Task {
+    protected String description;
+    protected boolean isDone;
```
Good naming on boolean variable!

desmondyst commented on src/main/java/luffy/Task.java:
```diff
@@ -0,0 +1,48 @@
+    protected String description;
+    protected boolean isDone;
+    protected String type;
+    protected String by;
+    protected String at;
```
Should this be extracted out into the relevant classes that `extends Task`? 🤔

desmondyst commented on src/main/java/luffy/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @author Silas Tay A0233425M
+ */
+public class Deadline extends Task{
```
```suggestion
public class Deadline extends Task {
```

desmondyst commented on src/main/java/luffy/Parser.java:
```diff
@@ -0,0 +1,118 @@
+     * @param s User input line
+     * @param tasks Tasklist of Luffy instance
+     */
+    public void parse(String s, TaskList tasks) {
```
This `parse` method seems to form the main bulk of the logic that handles the user's command. Perhaps you can think of creating `Command` classes and extract the logic into each corresponding `Command` class. E.g. `DeadlineCommand` class.

desmondyst reviewed (commented):
Majority of the code LGTM. Just a few nits to fix. However, the logic of the program seems to be nested in conditionals, which can get messy. Perhaps, you can try to separate the logic into smaller classes (e.g. the logic in `Parser` class). Other than that, good effort! 👍

SilasTSL committed:
Basic GUI without functionality

SilasTSL committed:
GUI with functionality

SilasTSL committed:
Add Assertion

Currently no assertions were used, assumptions not checked

Added Java assert to Tasklist delete method

To check if task being deleted is not null

SilasTSL committed:
Fix Parser Code Quality

SilasTSL committed:
Fix Code Quality

SilasTSL committed:
Revert "Fix Code Quality"

This reverts commit abefa9069c2bf4cbea21a8920906d0457e319295.

SilasTSL committed:
Merge branch 'branch-A-CodeQuality'

SilasTSL committed:
Add Deadlines feature

SilasTSL committed:
List command now responds to empty tasklist

SilasTSL committed:
Update User Guide

SilasTSL committed:
Fix localDate bug

SilasTSL committed:
Fix build.gradle file

SilasTSL committed:
Dummy commit

SilasTSL committed:
Merge pull request #2 from SilasTSL/branch-MergePR

Dummy commit

SilasTSL committed:
Fix CheckStyle

SilasTSL committed:
Changed README files
